src/main/java/io/cdap/plugin/ariba/source/AribaServices.java

```diff
@@ -332,51 +568,270 @@ private AribaResponseContainer aribaResponse(Response res) throws IOException {
+      .addHeader(ACCEPT, MediaType.APPLICATION_JSON)
+      .build();
+  }
+
+  /**
+   * @param req request
+   * @return Response
+   * @throws AribaException AribaException
+   * @throws InterruptedException InterruptedException
+   * @throws IOException IOException
+   */
+  public Response executeRequest(Request req) throws AribaException, InterruptedException, IOException {
+    OkHttpClient enhancedOkHttpClient = getConfiguredClient().build();
+    Response response = enhancedOkHttpClient.newCall(req).execute();
+    if (response.code() != HttpURLConnection.HTTP_OK && AribaUtil.isNullOrEmpty(response.message())) {
+      AribaResponseContainer responseContainer = aribaResponse(response);
+      InputStream responseStream = responseContainer.getResponseBody();
       JsonNode jsonNode = objectMapper.readTree(responseStream);
+      String errMsg = jsonNode.get(ResourceConstants.MESSAGE).asText() != null
+        ? jsonNode.get(ResourceConstants.MESSAGE).asText() :
+        ResourceConstants.ERR_NOT_FOUND.getMsgForKey();
+      throw new AribaException(errMsg, response.code());
+    }
+    if (!isApiLimitExhausted(response)) {
+      return response;
+    } else {
+      return enhancedOkHttpClient.newCall(req).execute();
+    }
+  }
 
-      HttpUrl.Builder builder = metaBuilder(jsonNode.get(DOCUMENT_TYPE).asText());
-      return builder.build().url();
+  /**
+   * @param jobId Ariba Job Id
+   * @return JsonNode
+   */
+  public JsonNode fetchData(String jobId, String fileName)
+    throws IOException, InterruptedException, AribaException {
+
+    OkHttpClient enhancedOkHttpClient = getConfiguredClient().build();
+    HttpUrl.Builder zipUrl = zipBuilder(jobId, fileName);
+    Response zipResponse;
+    do {
+      zipResponse = enhancedOkHttpClient
+        .newCall(fetchZipFileData(zipUrl.build().url(), getAccessToken())).execute();
+      if (zipResponse.code() == ResourceConstants.API_LIMIT_EXCEED) {
+        isApiLimitExhausted(zipResponse);
+      }
+    } while (zipResponse.code() == ResourceConstants.API_LIMIT_EXCEED);
+
+    LOG.info("Fetch Data Response Code is: {} for Job Id: {} , and File: {}"
+      , zipResponse.code(), jobId, fileName);
+
+    AribaResponseContainer responseContainer = tokenResponse(zipResponse);
+    try (InputStream responseStream = responseContainer.getResponseBody();
+         ZipInputStream zis = new ZipInputStream(Objects.requireNonNull(responseStream))) {
+      zis.getNextEntry();
+      return objectMapper.readTree(zis);
     }
   }
 
   /**
-   * Build URL to make call to view Template in order to fetch document type
+   * checking for api limits
    *
-   * @return HttpUrl.Builder
+   * @param res Request
+   * @return boolean
+   * @throws InterruptedException InterruptedException
+   * @throws AribaException AribaException
    */
-  private HttpUrl.Builder templateBuilder() {
+  public boolean isApiLimitExhausted(Response res) throws InterruptedException, AribaException {
```

Comment:
isApiLimitsExhausted() should just check if rate limit is exceeded and return true/false. It shouldn't throw exceptions, sleep, check for other error conditions etc. All of that logic belong in the caller. You can create other helper methods for handling those cases if needed

Reply:
Fixed